File: src/main/java/me/tbg/match/bot/MatchFinishListener.java
package me.tbg.match.bot;

import java.awt.Color;
import java.time.Instant;
import java.util.Comparator;
import java.util.Map;
import java.util.stream.Collectors;
import net.dv8tion.jda.api.EmbedBuilder;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import tc.oc.pgm.api.map.MapInfo;
import tc.oc.pgm.api.match.Match;
import tc.oc.pgm.api.match.event.MatchFinishEvent;
import tc.oc.pgm.api.party.Competitor;
import tc.oc.pgm.score.ScoreMatchModule;
import tc.oc.pgm.teams.TeamMatchModule;

public class MatchFinishListener implements Listener {

    private final DiscordBot bot;

    public MatchFinishListener(DiscordBot bot) {
        this.bot = bot;
    }

    @EventHandler
    public void onMatchFinish(MatchFinishEvent event) {
        Match match = event.getMatch();
        MapInfo map = match.getMap();
        ScoreMatchModule scoreModule = match.getModule(ScoreMatchModule.class);
        TeamMatchModule teamModule = match.getModule(TeamMatchModule.class);

        String winner = getWinner(event);
        Color winnerColor = getWinnerColor(event);

        EmbedBuilder matchFinishEmbed =
                createMatchFinishEmbed(match, map, winner, winnerColor, scoreModule, teamModule);

        bot.editMatchEmbed(Long.parseLong(match.getId()), matchFinishEmbed);
    }

    private String getWinner(MatchFinishEvent event) {
        if (event.getWinners().size() == 1) {
            return event.getWinners().iterator().next().getNameLegacy();
        } else if (event.getWinners().isEmpty()) {
            return "_No winner_";
        } else {
            return "Tie";
        }
    }

    private Color getWinnerColor(MatchFinishEvent event) {
        if (event.getWinners().size() == 1) {
            return new Color(event.getWinners().iterator().next().getFullColor().asRGB());
        } else {
            return Color.RED;
        }
    }

    private EmbedBuilder createMatchFinishEmbed(
            Match match,
            MapInfo map,
            String winner,
            Color winnerColor,
            ScoreMatchModule scoreModule,
            TeamMatchModule teamModule) {
        EmbedBuilder embed = new EmbedBuilder()
                .setColor(winnerColor.getRGB())
                .setTitle("Match #" + match.getId() + " has finished!")
                .setDescription("Started at <t:" + bot.getMatchStartTimestamp(Long.parseLong(match.getId()))
                        + ":f> with **"
                        + bot.getMatchStartPlayers(Long.parseLong(match.getId()))
                        + (bot.getMatchStartPlayers(Long.parseLong(match.getId())) == 1 ? " player" : " players")
                        + "**  and finished at <t:" + Instant.now().getEpochSecond() + ":f> with **"
                        + match.getPlayers().size() + (match.getPlayers().size() == 1 ? " player" : " players")
                        + "** online.")
                .addField("Winner", winner, true)
                .addField("Time", bot.parseDuration(match.getDuration()), true);

        if (scoreModule != null) {
            if (teamModule != null) {
                Map<String, Integer> teamScores = match.getCompetitors().stream()
                        .collect(Collectors.toMap(Competitor::getNameLegacy, team -> (int) scoreModule.getScore(team)));
                embed.addField("Scores", formatScores(teamScores), true);
            } else {
                Map<String, Integer> playerScores = match.getCompetitors().stream()
                        .collect(Collectors.toMap(
                                Competitor::getNameLegacy, player -> (int) scoreModule.getScore(player)));
                embed.addField("Podium", formatPodium(playerScores), true);
            }
        } else {
            embed.addField("\u200E", "\u200E", false);
        }

        embed.addField("Map", map.getName(), true)
                .addField("Version", map.getVersion().toString(), true)
                .addField("Gamemodes", bot.getMapGamemodes(match).toUpperCase(), true)
                .addField("Participants", String.valueOf(match.getParticipants().size()), true)
                .addField("Observers", String.valueOf(match.getDefaultParty().getPlayers().size()), true)
                .addField("Staff", String.valueOf(bot.getOnlineStaffCount(match)), true)
                .setFooter("Map tags: " + map.getTags().toString());

        bot.setEmbedThumbnail(map, embed, bot);

        return embed;
    }

    private String formatScores(Map<String, Integer> scores) {
        return scores.entrySet().stream()
                .map(e -> e.getKey() + ": " + e.getValue() + " points")
                .collect(Collectors.joining("\n"));
    }

    private String formatPodium(Map<String, Integer> scores) {
        return scores.entrySet().stream()
                .sorted(Map.Entry.comparingByValue(Comparator.reverseOrder()))
                .limit(3)
                .map(e -> e.getKey() + ": " + e.getValue() + " points")
                .collect(Collectors.joining("\n"));
    }
}

File: src/main/java/me/tbg/match/bot/MatchFinishListener.kt
package me.tbg.match.bot

import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import tc.oc.pgm.api.match.event.MatchFinishEvent
import tc.oc.pgm.score.ScoreMatchModule
import tc.oc.pgm.teams.TeamMatchModule
import org.javacord.api.entity.message.embed.EmbedBuilder
import java.awt.Color
import java.time.Instant
import java.util.Comparator
import java.util.HashMap
import java.util.stream.Collectors
import java.util.Locale

class MatchFinishListener(private val bot: DiscordBot) : Listener {
    @EventHandler
    fun onMatchFinish(event: MatchFinishEvent) {
        val match = event.match
        val map = match.map
        val scoreModule = match.getModule(ScoreMatchModule::class.java)
        val teamModule = match.getModule(TeamMatchModule::class.java)
        val teams = match.competitors
        var winner = ""
        var winnerColor: Color? = null
        for (competitor in teams) {
            if (event.winners.contains(competitor)) {
                if (event.winners.size == 1) {
                    winner = competitor.nameLegacy
                    winnerColor = Color(competitor.fullColor.asRGB())
                } else {
                    winner = "Tie"
                    winnerColor = Color.RED
                }
            }
        }
        val matchInfo = EmbedBuilder()
            .setColor(winnerColor)
            .setTitle("Match #" + match.id + " has finished!")
            .setThumbnail(bot.getMapImageUrl(map))
            .setDescription(
                "Finished at <t:"
                        + Instant.now().epochSecond
                        + ":f> with **"
                        + match.players.size
                        + (if (match.players.size == 1) " player" else " players")
                        + "** online."
            )
            .addInlineField("Winner", winner.ifEmpty { "_No winner_" })
            .addInlineField("Time", bot.parseDuration(match.duration))
        if (scoreModule != null) {
            if (teamModule != null) {
                val teamScores: MutableMap<String, Int> = HashMap()
                for (team in teams) {
                    teamScores[team.nameLegacy] = scoreModule.getScore(team).toInt()
                }
                matchInfo.addInlineField(
                    "Scores",
                    teamScores.entries.stream()
                        .map { (key, value): Map.Entry<String, Int> -> "$key: $value points" }
                        .collect(Collectors.joining("\n")))
            } else {
                val playerScores: MutableMap<String, Int> = HashMap()
                for (player in match.competitors) {
                    playerScores[player.nameLegacy] = scoreModule.getScore(player).toInt()
                }
                matchInfo.addInlineField(
                    "Podium",
                    playerScores.entries.stream()
                        .sorted(java.util.Map.Entry.comparingByValue(Comparator.reverseOrder()))
                        .limit(3)
                        .map { (key, value): Map.Entry<String, Int> -> "$key: $value points" }
                        .collect(Collectors.joining("\n")))
            }
        } else {
            matchInfo.addInlineField("\u200E", "\u200E")
        }
        matchInfo
            .addInlineField("Map", map.name)
            .addInlineField("Version", map.version.toString())
            .addInlineField("Gamemodes", bot.getMapGamemodes(match).uppercase(Locale.getDefault()))
            .addInlineField("Participants", match.participants.size.toString())
            .addInlineField("Observers", match.defaultParty.players.size.toString())
            .addInlineField("Staff", bot.getOnlineStaffCount(match).toString())
            .setFooter("Map tags: " + map.tags.toString())
        bot.sendMatchEmbed(matchInfo, match)
    }
}